### PR TITLE
LMBQ-113: During element screenshot, add 1px tolerance

### PR DESCRIPTION
**Pull request by 0liver**
[LMBQ-113](https://lombiq.atlassian.net/browse/LMBQ-113)

[LMBQ-113]: https://lombiq.atlassian.net/browse/LMBQ-113?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Lombiq.Tests.UI/Extensions/ScreenshotUITestContextExtensions.cs [hunk 1]
```diff
@@ -105,27 +105,48 @@ public static Image TakeElementScreenshot(this UITestContext context, IWebElemen
...
-            throw new InvalidOperationException(
-                "The captured screenshot size is smaller then the size required by the selected element. This can occur"
-                + " if there was an unsuccessful scrolling operation while capturing page parts."
-                + $"Captured size: {screenshot.Width.ToTechnicalString()} x {screenshot.Height.ToTechnicalString()}. "
-                + $"Required size: {elementAbsoluteSize.Width.ToTechnicalString()} x "
-                + $"{elementAbsoluteSize.Height.ToTechnicalString()}.");
+            // A difference of 1px can occur when both element.Location and element.Size were rounded to the next
+            // integer from exactly 0.5px, e.g. 212.5px + 287.5px = 500px in the browser, but due to both Size and Point
+            // using int for their coordinates, this will become 213px + 288px = 501px, which will be caught as an error
+            // here. In that case, we keep the elementSize as is, but reduce the Location coordinate by 1 so that cropping
```
Comment (DemeSzabolcs) [inline Lombiq.Tests.UI/Extensions/ScreenshotUITestContextExtensions.cs]: Long line. 120+

File: Lombiq.Tests.UI/Extensions/ScreenshotUITestContextExtensions.cs [hunk 2]
```diff
@@ -105,27 +105,48 @@ public static Image TakeElementScreenshot(this UITestContext context, IWebElemen
...
+                    "The captured screenshot size is smaller then the size required by the selected element. This can"
+                    + " occur if there was an unsuccessful scrolling operation while capturing page parts."
+                    + $" Captured size: {AsDimensions(screenshot)}. Expected size: {AsDimensions(expectedSize)}.");
+            }
         }
 
-        var bounds = new Rectangle(element.Location.X, element.Location.Y, element.Size.Width, element.Size.Height);
-        return screenshot.Clone(ctx => ctx.Crop(bounds));
+        var cropRectangle = new Rectangle(elementLocation.X, elementLocation.Y, elementSize.Width, elementSize.Height);
+        return screenshot.Clone(ctx => ctx.Crop(cropRectangle));
```
Comment (DemeSzabolcs) [inline Lombiq.Tests.UI/Extensions/ScreenshotUITestContextExtensions.cs]: I think changing `ctx` to something more descriptive would be good too.

File: Lombiq.Tests.UI/Extensions/ScreenshotUITestContextExtensions.cs [hunk 3]
```diff
@@ -105,27 +105,48 @@ public static Image TakeElementScreenshot(this UITestContext context, IWebElemen
...
     /// Takes a screenshot of an element region only.
     /// </summary>
     public static Image TakeElementScreenshot(this UITestContext context, By elementSelector) =>
         context.TakeElementScreenshot(context.Get(elementSelector));
+
+    private static string AsDimensions(IImageInfo image) =>
+        $"{image.Width.ToTechnicalString()} x {image.Height.ToTechnicalString()}";
+
+    private static string AsDimensions(Size size) =>
+        $"{size.Width.ToTechnicalString()} x {size.Height.ToTechnicalString()}";
```
Comment (DemeSzabolcs) [inline Lombiq.Tests.UI/Extensions/ScreenshotUITestContextExtensions.cs:151]: Both of these functions are only used in one place.

File: Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs [hunk 4]
```diff
@@ -21,6 +21,40 @@ public enum Browser
...
+
+    public static readonly Action<IEnumerable<LogEntry>> AssertBrowserLogIsEmpty =
+        logEntries => logEntries.ShouldNotContain(
+            logEntry => IsValidBrowserLogEntry(logEntry),
+            logEntries.Where(IsValidBrowserLogEntry).ToFormattedString());
+
+    public static readonly Func<LogEntry, bool> IsValidBrowserLogEntry =
+        logEntry =>
+            logEntry.Level >= LogLevel.Warning &&
+            // HTML imports are somehow used by Selenium or something but this deprecation notice is always there for every
```
Comment (DemeSzabolcs) [inline Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs]: Long comment. 120+ characters.

**Review by DemeSzabolcs — CHANGES_REQUESTED** [review]

File: Lombiq.Tests.UI/Extensions/ScreenshotUITestContextExtensions.cs (hunk 3 above)
Comment (0liver) [reply Lombiq.Tests.UI/Extensions/ScreenshotUITestContextExtensions.cs:151]: That is correct. They help keep the error message string in line 133 readable. I think this adds enough value to justify their existence.

**Review by DemeSzabolcs — APPROVED** [review]

**Pull request merged**